### ISSUE TITLE
Deliver standard dictionary with the app

### ISSUE THREAD

**Issue by tuxor1337** [tuxor1337/firedict#5]
One user on the Firefox Marketplace proposed to ship FireDict along with a standard dictionary (say english-english). Even comparably small dictionaries would certainly blow up the app's size unnecessarily. But I still get the point: Maybe it would be convenient to have a way to download the first dictionary from inside of the app. However, it seems to me like this will not work without additional permissions (at least sdcard write access).

**Comment by tuxor1337** [comment]
I don't want to add more permissions to FireDict and I don't want to have a bigger overall app size.

The best solution that I can't think of is: Writing a separate app (e.g. `FireDict Downloader`, it should be available from the Firefox Marketplace) that has sdcard write access and lets the user choose dictionaries from the list on http://tuxor1337.github.io/firedict/dictionaries.html. The dictionaries are downloaded by the app and unzipped to the `dictdata` directory so that they can be accessed immediately by FireDict.

**Comment by knezi** [comment]
Maybe we could have a link to this app (rather than marketplace) and download the dictionaries just on online web page.
EDIT: we need privileged permission, so we still have to install the app. Is there any way how to uninstall the app? Then we could make all almost automatically.

**Comment by tuxor1337** [comment]
Do I understand correctly that you want to have an app that downloads dictionaries and then uninstalls itself automatically?

**Comment by knezi** [comment]
Yeah I see. I overcomplicated it. I was thinking of way how to do it in the simplest way. So I was thinking about some auto install (with confirmation) and then uninstall itself (because the user needs to use it just once). But it sounds weird.

**Comment by tuxor1337** [comment]
It doesn't sound that weird actually. On Android there is the app "Document Viewer" and another app that is called "Document Viewer Font Pack". When a user launches "Document Viewer Font Pack" the app copies certain fonts to the local storage and then it uninstalls itself automatically.

The new part in your suggestion is that you want to automatically install the app "FireDict Downloader". This should be possible using `mozApps.installPackage()` (https://developer.mozilla.org/en-US/Marketplace/Options/Self_publishing#Self-publishing_packaged_apps) from within FireDict.

Unfortunately, I don't see a way to autoremove an app on Firefox OS.

**Comment by knezi** [comment]
That's great. We can do it in this way. Don't even have to make an app on marketplace, just host it on Github.
Maybe we could have uninstall the app just on user's demand, because we could provide the update of the dicts (I don't know about other languages, but the czech ones are updated quite often).
I am thinking about following:
* it could be app without icon, so it won't annoy the user (but how then uninstall the app? It is not so straightforward to go to the setting)
* Rewrite list of dictionaries into txt file and by make generate json file and HTML, so that it is more easily accessible for the download app
* implement more sofisticated operations on dictionaries (well, renaming is already implemented in dict app, but what about updating and deleting dicts)

The source code should have the same structure as FireDict, so there would be all the libraries. But since they are two distinc apps, there is no possibility how to share the libs, isn't it?
If you are not against, I will make a testing branch and try it.

**Comment by tuxor1337** [comment]
Of course, you may try it. But note the following remarks about your brainstorming:
* You have to upload this app to the marketplace because it needs to do cross-origin requests when downloading the dictionaries. And cross-origin requests are only allowed for privileged apps having the `systemXHR` permission.
* As I said, the app can't be uninstalled from within the app. Neither automatically nor on the user's demand. There's simply no API in Firefox OS to do so.
* There are no apps on Firefox OS that don't have icons.
* Certainly, we could base the app on AngularJS. But as you said, there's no way to have something like shared libraries between FireDict and the Downloader app.

**Comment by knezi** [comment]
* is it possible to install it automatically app even on Marketplace? If it is, then why not publish it on Marketplace.
* I meant demand like classical uninstall.
* Oh, you are right. I was thinking about copy keyboard, but it is just keyboard.
* It is separate app, but still, should it be in the same repository?

**Comment by tuxor1337** [comment]
Regarding your first point: Yes, it should be possible to trigger the installation of the helper-app from within FireDict even if the helper-app is installed from the Marketplace.

Regarding your last point: Actually, I prefer to have subprojects organized in separate git repositories.
